File: src/commands/utils/nyaa-magnet.ts
import { Constants, EmbedOptions } from "eris";
import { TofuCommand } from "@/core/command";
import { colors } from "@/utils/colors";
import { emojis } from "@/utils/emojis";
import { ErisUtils } from "@/utils/eris";
import { Nyaa, NyaaMagnetResult } from "@/utils/nyaa";

export const nyaaMagnetCommand: TofuCommand = {
    config: {
        name: "nyaa-magnet",
        description: "Fetches magnet link from Nyaa.si.",
        options: [
            {
                name: "id",
                description: "Nyaa ID.",
                type: Constants.ApplicationCommandOptionTypes.STRING,
                required: true,
            },
        ],
    },
    invoke: async (_, interaction) => {
        const id = ErisUtils.getCommandInteractionStringOptionValue(
            interaction,
            "id",
        );
        if (!id) {
            return {
                message: ErisUtils.failureMessage(
                    "You did not provide a value for `id`.",
                ),
            };
        }
        const result = await Nyaa.magnet(id);
        return {
            message: {
                embeds: [createEmbed(result)],
            },
        };
    },
};

function createEmbed(result: NyaaMagnetResult) {
    const embed: EmbedOptions = {
        author: {
            name: result.title,
            icon_url: Nyaa.faviconUrl,
            url: result.url,
        },
        fields: [
            {
                name: `${emojis.magnet} Magnet`,
                value: `\`\`\`${result.magnet}\`\`\``,
            },
        ],
        color: colors.nyaaBlue,
    };
    return embed;
}

File: internal/commands/nyaa_magnet.go
package commands

import (
	"fmt"
	"log/slog"
	"strings"

	"github.com/disgoorg/disgo/discord"
	"github.com/disgoorg/disgo/events"
	"me.zyrouge.tofu/internal/core"
	"me.zyrouge.tofu/internal/utils"
	"me.zyrouge.tofu/internal/utils/nyaa"
)

func NewTofuNyaaMagnetCommand() core.TofuCommand {
	return core.TofuCommand{
		Config: discord.SlashCommandCreate{
			Name:        "nyaa-magnet",
			Description: "Fetches magnet link from Nyaa.si.",
			Contexts: []discord.InteractionContextType{
				discord.InteractionContextTypeGuild,
			},
			Options: []discord.ApplicationCommandOption{
				discord.ApplicationCommandOptionString{
					Name:        "id",
					Description: "Nyaa ID.",
					Required:    true,
				},
			},
		},
		Invoke: func(tofu *core.Tofu, event *events.ApplicationCommandInteractionCreate) discord.InteractionResponseData {
			data := event.SlashCommandInteractionData()
			id := strings.TrimSpace(data.String("id"))
			if id == "" {
				return discord.NewMessageCreateBuilder().
					SetContent(utils.FailureMessage("You did not provide a value for `id`.")).
					Build()
			}
			result, err := nyaa.NyaaMagnet(id)
			if err != nil {
				slog.Error("nyaa magnet fetch failed: " + err.Error())
				return discord.NewMessageCreateBuilder().
					SetContent(utils.FailureMessage("Nyaa magnet fetch failed unknowingly.")).
					Build()
			}
			embed := discord.NewEmbedBuilder().
				SetAuthor(result.Title, result.Url, nyaa.NyaaFaviconUrl).
				SetColor(utils.ColorNyaaBlue).
				AddField(
					fmt.Sprintf("%s Magent", utils.EmojiMagnet),
					fmt.Sprintf("```%s```", result.Magnet),
					false,
				)
			return discord.NewMessageCreateBuilder().
				SetEmbeds(embed.Build()).
				Build()
		},
	}
}
